Allow to set Shared Memory size (shmSize) for GenericContainer

fzakaria opened the pull request:
Add the necessary code so that `shmSize` can be set on a GenericContainer.

This is especially useful for when starting Oracle docker images because they need at least 1GB of shared memory space.

> [docker run ... --shm-size=1g oracle/database:11.2.0.2-xe](https://github.com/oracle/docker-images/blob/master/OracleDatabase/SingleInstance/FAQ.md)

Fixes #952

rnorth commented:
Thanks @fzakaria.

Looks like CircleCI had an unrelated glitch downloading the gradle wrapper so I've restarted the build.

fzakaria commented:
Thanks @rnorth  -- looks to have been resolved.

fzakaria commented on core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java:
```diff
@@ -124,6 +126,13 @@ public static void setupContent() throws FileNotFoundException {
+     * Redis with shared memory
+     */
+    @ClassRule
+    public static GenericContainer redisWithSharedMemory = new GenericContainer("redis:3.0.2")
```
@rnorth this is following the same convention in the rest of the document -- it's not clear why Codacy is complaining.

fzakaria commented:
I updated the PR to have _shmSize_ be simply bytes instead of 'megabytes' to avoid mebibyte vs megabyte confusion.

fzakaria commented:
Looks like travis-ci had a flaky failure pulling an artifact -- could someone restart.
I believe this PR is ready to merge; thanks.

kiview commented on core/src/main/java/org/testcontainers/containers/GenericContainer.java:
```diff
@@ -21,6 +22,7 @@
 import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
 import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
 import org.apache.commons.compress.utils.IOUtils;
+import org.apache.commons.io.FileUtils;
```
Where did this import come from?

kiview commented on core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java:
```diff
@@ -124,6 +126,13 @@ public static void setupContent() throws FileNotFoundException {
+    /**
+     * Redis with shared memory
+     */
+    @ClassRule
```
I think using the `@ClassRule` for all containers in this test class is quite bad and we should instead start the containers in the test itself. It's not your fault of course and should not be part of this PR, just a heads up for the others.

fzakaria replied:
Moved to a try-with-resources ; also made it busybox to lighten the container that starts up.

kiview commented on core/src/main/java/org/testcontainers/containers/GenericContainer.java:
```diff
@@ -252,7 +261,14 @@ private void tryStart(Profiler profiler) {
             profiler.start("Create container");
-            CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
+
+            HostConfig hostConfig = new HostConfig();
```
I imagine we will continue adding support for more host config parameters, so I would prefer the `HostConfig` initialization in a private method, like `buildHostConfig()`.

kiview commented on core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java:
```diff
@@ -375,4 +384,13 @@ public void addExposedPortAfterWithExposedPortsTest() {
+
+    @Test
+    public void sharedMemorySetTest() {
+        assertEquals("Shared memory is not set", redisWithSharedMemory.getShmSize(), 1024 * FileUtils.ONE_MB);
```
I would remove this assert.

fzakaria commented:
Thanks for feedback @kiview  -- I've addressed them with latest update.
- moved to busybox + try-with-resources in the test
- moved HostConfig to a `buildHostConfig`
- removed extra assert

kiview commented on core/src/main/java/org/testcontainers/containers/GenericContainer.java:
```diff
@@ -1157,6 +1181,16 @@ public SELF withCreateContainerCmdModifier(Consumer<CreateContainerCmd> modifier
 
+    /**
+     * Size of /dev/shm
+     * @param bytes The number of megabybytestes to assign the shared memory. Null to not set any.
```
```suggestion
     * @param bytes The number of megabytes to assign the shared memory. Null to not set any.
```

Doesn't `null` mean we will use the Docker default of 64MB?

kiview commented on core/src/main/java/org/testcontainers/containers/GenericContainer.java:
```diff
@@ -1157,6 +1181,16 @@ public SELF withCreateContainerCmdModifier(Consumer<CreateContainerCmd> modifier
 
+    /**
+     * Size of /dev/shm
+     * @param bytes The number of megabytes to assign the shared memory. If null, it will apply the Docker default which is 64 KB.
```
```suggestion
     * @param bytes The number of megabytes to assign the shared memory. If null, it will apply the Docker default which is 64 MB.
```

Isn't it MB?

fzakaria replied:
Darn it! I went off the Oracle docs but you are write.

http://www.tin.org/bin/man.cgi?section=8&topic=dockerd

```
    --default-shm-size=64MiB
         Set the daemon-wide default shm size for containers. Default is
       64MiB.
```

fzakaria commented:
@kiview What's next step to get this merged & published?

kiview commented:
Hey @fzakaria, sorry for letting this hanging, I think it's good to go.
I'll ping the others about final feedback and then we can merge.

bsideup commented on core/src/main/java/org/testcontainers/containers/GenericContainer.java:
```diff
@@ -252,7 +260,12 @@ private void tryStart(Profiler profiler) {
             profiler.start("Create container");
-            CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
+
+            HostConfig hostConfig = buildHostConfig();
```
please move it to `applyConfiguration`

fzakaria commented:
@bsideup made changes requested.

fzakaria commented:
Update on this ? Quite a lot of effort to get a helpful change in.
I might abandon the process soon.

fzakaria commented:
@kiview thanks. 
Sorry for the attitude -- other stuff that day was likely getting to me.

kiview commented:
Travis failure unrelated, restarted.

kiview commented:
@fzakaria merged, thanks a lot for the great PR as well as your patience :slightly_smiling_face:
